### PR TITLE
AC-888 Migrated Login Package from Java to Kotlin

### PR DESCRIPTION
**Pull request by shivamsawlani02**
## Description of what I changed
Migrated Login Package which consisted of five files(Activity, Contract, Presenter, Fragment, and Location Adapter) from java to kotlin.

## Issue I worked on
JIRA Issue: https://issues.openmrs.org/browse/AC-888

## Checklist: I completed these to help reviewers :)

- [x] My pull request only contains **ONE single commit**
(the number above, next to the 'Commits' tab is 1).

- [ ] I have **added tests** to cover my changes. (If you refactored
existing code that was well tested you do not have to add tests)

- [x] All new and existing **tests passed**.

- [x] My pull request is **based on the latest changes** of the master branch.

### CODE REVIEW

**Comment by LuGO0** [comment]
@shivamsawlani02 did you check the travis?

**Comment by shivamsawlani02** [comment]
@LuGO0 yes I checked but could not configure why the travis build is failing.

**Comment by shivamsawlani02** [comment]
@LuGO0 the Travis CI build passed successfully now.

**Comment by sahuadarsh0** [comment]
👍🏼

File: openmrs-client/src/main/java/org/openmrs/mobile/activities/login/LocationArrayAdapter.kt [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+import android.widget.ArrayAdapter
+import android.widget.TextView
+
+class LocationArrayAdapter(context: Context?, objects: List<String?>?) : ArrayAdapter<String?>(context!!, R.layout.simple_spinner_item, objects!!) {
+    override fun getCount(): Int {
+        return super.getCount()
+    }
+
+    override fun getDropDownView(position: Int, convertView: View?, parent: ViewGroup?): View {
+        val v: View
```
Comment (rishabh-997) [inline openmrs-client/src/main/java/org/openmrs/mobile/activities/login/LocationArrayAdapter.kt:29]: use more informative naming convention

File: openmrs-client/src/main/java/org/openmrs/mobile/activities/login/LocationArrayAdapter.kt [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        if (position == 0) {
+            val tv = TextView(context)
```
Comment (rishabh-997) [inline openmrs-client/src/main/java/org/openmrs/mobile/activities/login/LocationArrayAdapter.kt:31]: same here

File: openmrs-client/src/main/java/org/openmrs/mobile/activities/login/LoginFragment.kt [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+            } else if (binding!!.loginUsernameField.text.toString() == "") {
+                binding!!.textInputLayoutUsername.hint = Html.fromHtml(getString(R.string.login_username_hint) + getString(R.string.req_star))
+                binding!!.textInputLayoutUsername.isHintAnimationEnabled = true
+            }
+        }
+        binding!!.loginPasswordField.onFocusChangeListener = OnFocusChangeListener { view: View?, hasFocus: Boolean ->
+            if (hasFocus) {
+                binding!!.textInputLayoutPassword.hint = Html.fromHtml(getString(R.string.login_password_hint))
+            } else if (binding!!.loginPasswordField.text.toString() == "") {
+                binding!!.textInputLayoutPassword.hint = Html.fromHtml(getString(R.string.login_password_hint) + getString(R.string.req_star))
```
Comment (rishabh-997) [inline openmrs-client/src/main/java/org/openmrs/mobile/activities/login/LoginFragment.kt:103]: use with(binding) {
/////
}

File: openmrs-client/src/main/java/org/openmrs/mobile/activities/login/LoginFragment.kt [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+                binding!!.textInputLayoutPassword.isHintAnimationEnabled = true
+            }
+        }
+        binding!!.loginButton.setOnClickListener { view: View? ->
+            mPresenter!!.login(binding!!.loginUsernameField.text.toString(),
+                    binding!!.loginPasswordField.text.toString(),
+                    binding!!.loginUrlField.text.toString(),
+                    initialUrl)
+        }
+        binding!!.forgotPass.setOnClickListener { view: View? -> startActivity(Intent(context, ContactUsActivity::class.java)) }
```
Comment (rishabh-997) [inline openmrs-client/src/main/java/org/openmrs/mobile/activities/login/LoginFragment.kt:113]: use with(binding) {
/////
}

File: openmrs-client/src/main/java/org/openmrs/mobile/activities/login/LoginFragment.kt [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+        binding!!.aboutUsTextView.setOnClickListener { view: View? -> openAboutPage() }
+    }
+
+    private fun initViewFields() {
+        binding!!.textInputLayoutPassword.hint = Html.fromHtml(getString(R.string.login_password_hint) + getString(R.string.req_star))
+        binding!!.textInputLayoutUsername.hint = Html.fromHtml(getString(R.string.login_username_hint) + getString(R.string.req_star))
```
Comment (rishabh-997) [inline openmrs-client/src/main/java/org/openmrs/mobile/activities/login/LoginFragment.kt:119]: use with(binding) {
/////
}

File: openmrs-client/src/main/java/org/openmrs/mobile/activities/login/LoginFragment.kt [hunk 6]
```diff
@@ -0,0 +1,335 @@
...
+        (this.activity as LoginActivity?)!!.createAndShowDialog(bundle, ApplicationConstants.DialogTAG.WARNING_LOST_DATA_DIALOG_TAG)
+    }
+
+    override fun showLoadingAnimation() {
+        binding!!.loginFormView.visibility = View.GONE
+        binding!!.loginLoading.visibility = View.VISIBLE
+    }
+
+    override fun hideLoadingAnimation() {
+        binding!!.loginFormView.visibility = View.VISIBLE
```
Comment (rishabh-997) [inline openmrs-client/src/main/java/org/openmrs/mobile/activities/login/LoginFragment.kt:176]: use with(binding) {
/////
}

File: openmrs-client/src/main/java/org/openmrs/mobile/activities/login/LoginFragment.kt [hunk 7]
```diff
@@ -0,0 +1,335 @@
...
+        bundle.textViewMessage = getString(R.string.warning_lost_data_dialog)
+        bundle.rightButtonText = getString(R.string.dialog_button_ok)
+        bundle.rightButtonAction = CustomFragmentDialog.OnClickAction.LOGIN
+        bundle.leftButtonText = getString(R.string.dialog_button_cancel)
+        bundle.leftButtonAction = CustomFragmentDialog.OnClickAction.DISMISS
+        (this.activity as LoginActivity?)!!.createAndShowDialog(bundle, ApplicationConstants.DialogTAG.WARNING_LOST_DATA_DIALOG_TAG)
+    }
+
+    override fun showLoadingAnimation() {
+        binding!!.loginFormView.visibility = View.GONE
```
Comment (rishabh-997) [inline openmrs-client/src/main/java/org/openmrs/mobile/activities/login/LoginFragment.kt:171]: use with(binding) {
/////
}

File: openmrs-client/src/main/java/org/openmrs/mobile/activities/login/LoginFragment.kt [hunk 8]
```diff
@@ -0,0 +1,335 @@
...
+    }
+
+    override fun showLocationLoadingAnimation() {
+        binding!!.loginButton.isEnabled = false
+        binding!!.locationLoadingProgressBar.visibility = View.VISIBLE
+    }
+
+    override fun hideUrlLoadingAnimation() {
+        binding!!.locationLoadingProgressBar.visibility = View.GONE
+        binding!!.loginLoading.visibility = View.GONE
```
Comment (rishabh-997) [inline openmrs-client/src/main/java/org/openmrs/mobile/activities/login/LoginFragment.kt:187]: similarly everwhere

**Review by rishabh-997 — CHANGES_REQUESTED** [review]
.

File: openmrs-client/src/main/java/org/openmrs/mobile/activities/login/LoginFragment.kt (hunk 3 above)
Comment (shivamsawlani02) [reply openmrs-client/src/main/java/org/openmrs/mobile/activities/login/LoginFragment.kt]: Wow! Had never used 'with' function before. Thanks for the info.

**Comment by shivamsawlani02** [comment]
@rishabh-997 made the changes. Please review.

File: openmrs-client/src/main/java/org/openmrs/mobile/activities/login/LoginFragment.kt (hunk 6 above)
Comment (sahuadarsh0) [reply openmrs-client/src/main/java/org/openmrs/mobile/activities/login/LoginFragment.kt]: also binding.apply { 
////
}
can be used

**Comment by LuGO0** [comment]
Closing this for now. please open the PR again after fixing the merge conflicts. thanks

**Pull request closed without merging**